[DebugInfo] Generate full debug info for generic structs

augusto2112 opened the pull request.

augusto2112 commented:
@swift-ci smoke test

slavapestov commented on lib/IRGen/IRGenDebugInfo.cpp:
```diff
@@ -1037,6 +1037,84 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
+      auto memberTy =
+          UnsubstitutedType->getTypeOfMember(IGM.getSwiftModule(), VD);
+      llvm::DIType *DITy = nullptr;
+      if (llvm::isa<swift::GenericTypeParamType>(memberTy)) {
```
Just `isa<GenericTypeParamType>` should work. However, I don't see what is special about generic parameter types here. You can also have a member type (T.A), wouldn't that be the same as a generic parameter?

augusto2112 replied:
You're right, this was incorrect.

slavapestov commented on lib/IRGen/IRGenDebugInfo.cpp:
```diff
@@ -1037,6 +1037,84 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
+    SmallVector<llvm::Metadata *, 16> Elements;
+    for (VarDecl *VD : Decl->getStoredProperties()) {
+      auto memberTy =
+          UnsubstitutedType->getTypeOfMember(IGM.getSwiftModule(), VD);
```
Isn't this just `VD->getInterfaceType()`?

augusto2112 replied:
It is! Thanks

slavapestov commented on lib/IRGen/IRGenDebugInfo.cpp:
```diff
@@ -1037,6 +1037,84 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
+  /// debug info generation doesn't attempt to emit the size and aligment of
+  /// the type, as those are all dependent on substituting the type parameters
+  /// in. It also doesn't attempt to emit the offset of the members for the same
+  /// reason.
```
This isn't strictly true if some of those type parameters are class-constrained, eg

```
protocol P {
  associatedtype A: AnyObject
}

struct Foo<T: P> {
  var x: T.A
}
```

augusto2112 replied:
Thanks, I'll update the comment to be more accurate.

adrian-prantl commented on lib/IRGen/DebugTypeInfo.h:
```diff
@@ -72,8 +72,9 @@ class DebugTypeInfo {
-  static DebugTypeInfo getForwardDecl(swift::Type Ty);
+  /// Create a forward declaration for a type whose size is unknown (forward
+  /// decls, unsubstituted generic type parameters)
+  static DebugTypeInfo getFromType(swift::Type Ty);
```
If this still created a forward declaration, I'd rather keep that in the function name?

adrian-prantl commented on lib/IRGen/DebugTypeInfo.h:
```diff
@@ -116,7 +117,10 @@ class DebugTypeInfo {
-  bool isForwardDecl() const { return FragmentStorageType == nullptr; }
+  bool isForwardDecl() const {
+    return FragmentStorageType == nullptr &&
+           !llvm::isa<swift::GenericTypeParamType>(Type);
```
Ah I see. So only generic types are not forward declarations?

augusto2112 replied:
Yes, but on second thought, maybe keeping generic type parameters types as forward declarations is better after all. I'll undo these changes.

adrian-prantl commented on lib/IRGen/IRGenDebugInfo.cpp:
```diff
@@ -1037,6 +1037,84 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
+      unsigned RuntimeLang, StringRef UniqueID) {
+    // FIXME: ideally, we'd like to emit this type with no size and alignment at
+    // all (instead of emitting them as 0). Fix this by changing DIBuilder to
+    // allow for struct types that have optional size and alignment.
```
From the DWARF perspective types with unknown size are forward declarations. Is the problem that you have is that you want a sizeless type that still has members?

augusto2112 replied:
Yes, what I meant was that a size such as `Pair<T, U>` doesn't have a size until the generic type parameters are substituted in, but we can still describe it in debug info as having two fields.

augusto2112 committed:
[DebugInfo] Generate full debug info for generic structs

Generate the full debug info for generic structs. The strategy is to
emit one full entry for the generic type with archetypes, and one
forward declaration per instantiation of the generic type.

For example, given:
```
struct Pair<T, U> {
  let t: T
  let u: U
}
let p1 = Pair<Int, Double>(t: 1, u: 4.2)
let p2 = Pair<String, Bool>(t: "Hello", u: true)
```

DebugInfo will have one entry for Pair<T, U> which includes descriptions
of its fields, one forward declaration of Pair<Int, Double> and one
forward declaration of Pair<String, Bool>. This information is enough
for the algorithms of RemoteMirrors to reconstruct type information for
the fully instantiated types.

augusto2112 commented:
@swift-ci test macOS

augusto2112 commented:
@swift-ci smoke test

augusto2112 commented:
@swift-ci smoke test

adrian-prantl commented on lib/IRGen/IRGenDebugInfo.cpp:
```diff
@@ -1037,6 +1046,48 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
     return DITy;
   }
 
+  /// Creates debug info for a generic struct with archetypes (eg.:
```
```suggestion
  /// Creates debug info for a generic struct with archetypes (e.g.:
```

adrian-prantl commented on lib/IRGen/IRGenDebugInfo.cpp:
```diff
@@ -1603,6 +1668,42 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
+        }
+        // Force the creation of the unsubstituted type, don't create it
+        // directly so it goes through all the caching/verification logic.
+        getOrCreateType(DbgTy);
```
What holds on to this type so it stays alive in the IR?

augusto2112 replied:
I thought `getOrCreateType` storing it in the type cache would be enough to keep it alive?

```
    DITypeCache.insert({DbgTy.getType(), llvm::TrackingMDNodeRef(DITy)});
```

augusto2112 replied:
@adrian-prantl the type was only being retained because I happened to be using one of its method in the test program (for example, its init). I changed this to always forcibly retain the type instead.

augusto2112 commented:
@swift-ci smoke test

augusto2112 commented:
@swift-ci smoke test Windows

augusto2112 commented:
@swift-ci smoke test Windows

augusto2112 commented:
@swift-ci smoke test

augusto2112 commented:
@swift-ci smoke test Windows
